### PR TITLE
feat: select entity/project on publish new

### PR DESCRIPTION
**Pull request by jamie-rasmussen**
When publishing a new board, allow the user to select entity and project.

Screenshot 2023-07-19 at 12 28 49 PM

Internal Jira: https://wandb.atlassian.net/browse/WB-14207 (partial)

### CODE REVIEW

File: weave-js/src/components/PagePanelComponents/PublishModal.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+          value: entName,
+          text: entName,
+        }));
+
+  const projectsNode = w.opEntityProjects({
+    entity: w.opRootEntity({
+      entityName: w.constString(entityName),
+    }),
+  });
+  const projectDataNode = w.opMap({
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/PublishModal.tsx:88]: This weave expression can be simplified to `w.opProjectName({project: projectsNode})`

File: weave-js/src/components/PagePanelComponents/PublishModal.tsx [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+  // Initialize the entity selector to first option (user entity)
+  useEffect(() => {
+    if (!userEntities.loading && entityOptions.length > 0) {
+      setEntityName(entityOptions[0].value as string);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [userEntities.loading]);
+
+  const projectNamesValue = useNodeValue(projectDataNode, {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/PublishModal.tsx:105]: would be nice to refactor out the query for projects for entity to the `query` file

File: weave-js/src/components/PagePanelComponents/persistenceStateMachine.ts [hunk 3]
```diff
@@ -272,15 +272,22 @@ export const useStateMachine = (
...
-            project_name: constString('weave'), // TODO: Allow user to specify name
+                : constNone(),
+            project_name:
+              actionOptions.projectName != null
+                ? constString(actionOptions.projectName)
+                : constString('weave'),
+            entity_name:
+              actionOptions.entityName != null
+                ? constString(actionOptions.entityName)
+                : constString(''),
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/persistenceStateMachine.ts:283]: should the default here be a constNone()?

File: weave-js/src/components/PagePanelComponents/persistenceStateMachine.ts [hunk 4]
```diff
@@ -272,15 +272,22 @@ export const useStateMachine = (
...
           },
           newRoot => {
             updateNode(newRoot);
           }
         );
       } else if (action === 'commit') {
-        await makeMutation(inputNode, 'merge', {}, newRoot => {
+        await makeMutation(inputNode, 'merge_artifact', {}, newRoot => {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/persistenceStateMachine.ts:290]: nice!!

**Review by tssweeney — APPROVED** [review]
HOT 🔥

File: weave-js/src/components/PagePanelComponents/persistenceStateMachine.ts (hunk 3 above)
Comment (jamie-rasmussen) [reply weave-js/src/components/PagePanelComponents/persistenceStateMachine.ts:283]: Probably yes - I think in the current code you can't actually hit this path

**Pull request merged**